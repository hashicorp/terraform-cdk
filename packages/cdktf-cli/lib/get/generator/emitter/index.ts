export * from './attributes-emitter'
export * from './resource-emitter'
export * from './struct-emitter'