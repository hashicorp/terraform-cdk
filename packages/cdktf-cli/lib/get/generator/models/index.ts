export * from "./attribute-model"
export * from "./attribute-type-model"
export * from "./resource-model"
export * from "./scope"
export * from "./struct"