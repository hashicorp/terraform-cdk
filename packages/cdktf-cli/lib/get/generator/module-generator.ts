import { CodeMaker, toCamelCase } from "codemaker";
import { Module, Submodule } from "../registry-client";

export class ModuleGenerator {
  constructor(private readonly code: CodeMaker, private readonly spec: Module) {
    this.code.indentation = 2;
    this.emitSubmodule(spec.root);
  }

  private emitSubmodule(spec: Submodule) {
    const source = `${this.spec.namespace}/${this.spec.name}/${this.spec.provider}`;
    const fileName = `modules/${source}.ts`;
    this.code.openFile(fileName);

    this.code.line(`// generated by cdktf get`);
    this.code.line(`// ${this.spec.id}/${spec.path}`);

    this.code.line(`import { TerraformModule } from 'cdktf';`);
    this.code.line(`import { Construct } from 'constructs';`);

    const baseName = this.code.toPascalCase(spec.name.replace(/-/g, '_'));
    const optionsType = `${baseName}Options`;

    this.code.openBlock(`export interface ${optionsType}`);
    for (const input of spec.inputs) {
      const optional = (input.required && (input.default === undefined)) ? '' : '?';
      this.code.line(`/**`);
      this.code.line(` * ${input.description}`);
      if (input.default) {
        this.code.line(` * @default ${input.default}`);
      }
      this.code.line(` */`);
      this.code.line(`readonly ${toCamelCase(input.name)}${optional}: ${parseType(input.type)};`);
    }
    this.code.closeBlock();

    this.code.openBlock(`export class ${baseName} extends TerraformModule`);

    this.code.line(`private readonly inputs: { [name: string]: any } = { }`);

    const allOptional = spec.inputs.find(x => x.required) ? '' : ' = {}';

    this.code.open(`public constructor(scope: Construct, id: string, options: ${optionsType}${allOptional}) {`);
    this.code.open(`super(scope, id, {`);
    this.code.line(`source: '${source}',`);
    this.code.line(`version: '${this.spec.version}',`);
    this.code.close(`});`);

    for (const input of spec.inputs) {
      const inputName = toCamelCase(input.name);
      this.code.line(`this.${inputName} = options.${inputName};`);
    }

    this.code.close(`}`); // ctor

    for (const input of spec.inputs) {
      const inputName = toCamelCase(input.name);
      const inputType = parseType(input.type) + ((input.required && (input.default === undefined)) ? '' : ' | undefined');
      this.code.openBlock(`public get ${inputName}(): ${inputType}`);
      this.code.line(`return this.inputs['${input.name}'] as ${inputType};`);
      this.code.closeBlock();

      this.code.openBlock(`public set ${inputName}(value: ${inputType})`);
      this.code.line(`this.inputs['${input.name}'] = value;`);
      this.code.closeBlock();
    }

    this.code.openBlock(`protected synthesizeAttributes()`);
    this.code.line(`return this.inputs;`);
    this.code.closeBlock();

    this.code.closeBlock(); // class

    this.code.closeFile(fileName);
  }

}

function parseType(type: string) {
  if (type === 'string') { return 'string'; }
  if (type === 'number') { return 'number'; }
  if (type === 'bool') { return 'boolean'; }
  if (type === 'list') { return 'string[]'; }
  if (type === 'map') { return '{ [key: string]: string }'; }
  if (type === 'any') { return 'any'; }

  const complexType = parseComplexType(type);
  if (complexType) { return complexType; }

  throw new Error(`unknown type ${type}`);
}

function parseComplexType(type: string): string | undefined {
  const complex = /^(object|list|map)\(([\s\S]+)\)/;
  const match = complex.exec(type);
  if (!match) {
    return undefined;
  }

  const [ , kind, innerType ] = match;

  if (kind === 'object') {
    return `any`;
  }

  if (kind === 'list') {
    return `${parseType(innerType)}[]`;
  }

  if (kind === 'map') {
    return `{ [key: string]: ${parseType(innerType) } }`;
  }

  throw new Error(`unexpected kind ${kind}`);
}