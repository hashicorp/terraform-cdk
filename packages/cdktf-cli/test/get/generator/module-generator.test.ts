import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { ModuleGenerator } from '../../../lib/get/generator/module-generator';
import { CodeMaker } from 'codemaker';

test('generate some modules', async () => {
  const code = new CodeMaker()
  const workdir = fs.mkdtempSync(path.join(os.tmpdir(), 'module-generator.test'));
  const spec = JSON.parse(fs.readFileSync(path.join(__dirname, 'fixtures', 'module-generator.test.fixture.json'), 'utf-8'));
  new ModuleGenerator(code, spec);
  await code.save(workdir);

  const output = fs.readFileSync(path.join(workdir, 'modules/terraform-aws-modules/eks/aws.ts'), 'utf-8');
  expect(output).toMatchSnapshot();
});