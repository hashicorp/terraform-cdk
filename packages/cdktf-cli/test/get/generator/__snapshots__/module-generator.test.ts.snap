// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate some modules 1`] = `
"// generated by cdktf get
// terraform-aws-modules/eks/aws/7.0.1/
import { TerraformModule } from 'cdktf';
import { Construct } from 'constructs';
export interface EksOptions {
  /**
   * Whether to attach the Amazon managed \`AmazonEKS_CNI_Policy\` IAM policy to the default worker IAM role. WARNING: If set \`false\` the permissions must be assigned to the \`aws-node\` DaemonSet pods via another method or nodes will not be able to join the cluster.
   * @default true
   */
  readonly attachWorkerCniPolicy?: boolean;
  /**
   * User defined workers role name.
   * @default \\"\\"
   */
  readonly workersRoleName?: string;
  /**
   * Additional policies to be added to workers
   * @default []
   */
  readonly workersAdditionalPolicies?: string[];
  /**
   * If provided, all workers will be attached to this security group. If not given, a security group will be created with necessary ingress/egress to work with the EKS cluster.
   * @default \\"\\"
   */
  readonly workerSecurityGroupId?: string;
  /**
   * Additional IAM users to add to the aws-auth configmap. See examples/basic/variables.tf for example format.
   * @default []
   */
  readonly mapUsers?: any[];
  /**
   * Additional IAM roles to add to the aws-auth configmap. See examples/basic/variables.tf for example format.
   * @default []
   */
  readonly mapRoles?: any[];
  /**
   * Whether to write the aws-auth configmap file.
   * @default true
   */
  readonly writeAwsAuthConfig?: boolean;
  /**
   * Whether to let the module manage worker IAM resources. If set to false, iam_instance_profile_name must be specified for workers.
   * @default true
   */
  readonly manageWorkerIamResources?: boolean;
  /**
   * Indicates whether or not the Amazon EKS public API server endpoint is enabled.
   * @default true
   */
  readonly clusterEndpointPublicAccess?: boolean;
  /**
   * Whether to create a security group for the workers or attach the workers to \`worker_security_group_id\`.
   * @default true
   */
  readonly workerCreateSecurityGroup?: boolean;
  /**
   * Default arguments passed to the authenticator command. Defaults to [token -i $cluster_name].
   * @default []
   */
  readonly kubeconfigAwsAuthenticatorCommandArgs?: string[];
  /**
   * Minimum port number from which pods will accept communication. Must be changed to a lower value if some pods in your cluster will expose a port lower than 1025 (e.g. 22, 80, or 443).
   * @default 1025
   */
  readonly workerSgIngressFromPort?: number;
  /**
   * The ID of the owner for the AMI to use for the AWS EKS Windows workers. Valid values are an AWS account ID, 'self' (the current account), or an AWS owner alias (e.g. 'amazon', 'aws-marketplace', 'microsoft').
   * @default \\"801119661308\\"
   */
  readonly workerAmiOwnerIdWindows?: string;
  /**
   * Name filter for AWS EKS worker AMI. If not provided, the latest official AMI for the specified 'cluster_version' is used.
   * @default \\"\\"
   */
  readonly workerAmiNameFilter?: string;
  /**
   * If provided, the EKS cluster will be attached to this security group. If not given, a security group will be created with necessary ingress/egress to work with the workers
   * @default \\"\\"
   */
  readonly clusterSecurityGroupId?: string;
  /**
   * Whether to attach the module managed cluster autoscaling iam policy to the default worker IAM role. This requires \`manage_worker_autoscaling_policy = true\`
   * @default true
   */
  readonly attachWorkerAutoscalingPolicy?: boolean;
  /**
   * Command to use to fetch AWS EKS credentials.
   * @default \\"aws-iam-authenticator\\"
   */
  readonly kubeconfigAwsAuthenticatorCommand?: string;
  /**
   * Name filter for AWS EKS Windows worker AMI. If not provided, the latest official AMI for the specified 'cluster_version' is used.
   * @default \\"\\"
   */
  readonly workerAmiNameFilterWindows?: string;
  /**
   * Override default values for target groups. See workers_group_defaults_defaults in local.tf for valid keys.
   * @default {}
   */
  readonly workersGroupDefaults?: any;
  /**
   * A list of maps defining worker group configurations to be defined using AWS Launch Configurations. See workers_group_defaults for valid keys.
   * @default []
   */
  readonly workerGroups?: any;
  /**
   * A map of tags to add to all resources.
   * @default {}
   */
  readonly tags?: { [key: string]: string };
  /**
   * Whether to let the module manage cluster IAM resources. If set to false, cluster_iam_role_name must be specified.
   * @default true
   */
  readonly manageClusterIamResources?: boolean;
  /**
   * Whether to create initial lifecycle hooks provided in worker groups.
   * @default false
   */
  readonly workerCreateInitialLifecycleHooks?: boolean;
  /**
   * Timeout value when deleting the EKS cluster.
   * @default \\"15m\\"
   */
  readonly clusterDeleteTimeout?: string;
  /**
   * Any additional arguments to pass to the authenticator such as the role to assume. e.g. [\\"-r\\", \\"MyEksRole\\"].
   * @default []
   */
  readonly kubeconfigAwsAuthenticatorAdditionalArgs?: string[];
  /**
   * The ID of the owner for the AMI to use for the AWS EKS workers. Valid values are an AWS account ID, 'self' (the current account), or an AWS owner alias (e.g. 'amazon', 'aws-marketplace', 'microsoft').
   * @default \\"602401143452\\"
   */
  readonly workerAmiOwnerId?: string;
  /**
   * A list of maps defining worker group configurations to be defined using AWS Launch Templates. See workers_group_defaults for valid keys.
   * @default []
   */
  readonly workerGroupsLaunchTemplate?: any;
  /**
   * If provided, all IAM roles will be created with this permissions boundary attached.
   */
  readonly permissionsBoundary?: string;
  /**
   * Whether to let the module manage the cluster autoscaling iam policy.
   * @default true
   */
  readonly manageWorkerAutoscalingPolicy?: boolean;
  /**
   * IAM role name for the cluster. Only applicable if manage_cluster_iam_resources is set to false.
   * @default \\"\\"
   */
  readonly clusterIamRoleName?: string;
  /**
   * Indicates whether or not the Amazon EKS private API server endpoint is enabled.
   * @default false
   */
  readonly clusterEndpointPrivateAccess?: boolean;
  /**
   * Whether to create a security group for the cluster or attach the cluster to \`cluster_security_group_id\`.
   * @default true
   */
  readonly clusterCreateSecurityGroup?: boolean;
  /**
   * VPC where the cluster and workers will be deployed.
   */
  readonly vpcId?: string;
  /**
   * A list of subnets to place the EKS cluster and workers within.
   */
  readonly subnets?: string[];
  /**
   * Whether to write a Kubectl config file containing the cluster configuration. Saved to \`config_output_path\`.
   * @default true
   */
  readonly writeKubeconfig?: boolean;
  /**
   * Kubernetes version to use for the EKS cluster.
   * @default \\"1.14\\"
   */
  readonly clusterVersion?: string;
  /**
   * A list of the desired control plane logging to enable. For more information, see Amazon EKS Control Plane Logging documentation (https://docs.aws.amazon.com/eks/latest/userguide/control-plane-logs.html)
   * @default []
   */
  readonly clusterEnabledLogTypes?: string[];
  /**
   * Timeout value when creating the EKS cluster.
   * @default \\"15m\\"
   */
  readonly clusterCreateTimeout?: string;
  /**
   * Override the default name used for items kubeconfig.
   * @default \\"\\"
   */
  readonly kubeconfigName?: string;
  /**
   * Environment variables that should be used when executing the authenticator. e.g. { AWS_PROFILE = \\"eks\\"}.
   * @default {}
   */
  readonly kubeconfigAwsAuthenticatorEnvVariables?: { [key: string]: string };
  /**
   * If provided, all IAM roles will be created on this path.
   * @default \\"/\\"
   */
  readonly iamPath?: string;
  /**
   * Command to run for local-exec resources. Must be a shell-style interpreter. If you are on Windows Git Bash is a good choice.
   * @default [
  \\"/bin/sh\\",
  \\"-c\\"
]
   */
  readonly localExecInterpreter?: string[];
  /**
   * A list of additional security group ids to attach to worker instances
   * @default []
   */
  readonly workerAdditionalSecurityGroupIds?: string[];
  /**
   * Additional AWS account numbers to add to the aws-auth configmap. See examples/basic/variables.tf for example format.
   * @default []
   */
  readonly mapAccounts?: string[];
  /**
   * Name of the EKS cluster. Also used as a prefix in names of related resources.
   */
  readonly clusterName?: string;
  /**
   * Whether to apply the aws-auth configmap file.
   * @default true
   */
  readonly manageAwsAuth?: string;
  /**
   * Where to save the Kubectl config file (if \`write_kubeconfig = true\`). Assumed to be a directory if the value ends with a forward slash \`/\`.
   * @default \\"./\\"
   */
  readonly configOutputPath?: string;
  /**
   * Number of days to retain log events. Default retention - 90 days.
   * @default 90
   */
  readonly clusterLogRetentionInDays?: number;
  /**
   * If a KMS Key ARN is set, this key will be used to encrypt the corresponding log group. Please be sure that the KMS Key has an appropriate key policy (https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/encrypt-log-data-kms.html)
   * @default \\"\\"
   */
  readonly clusterLogKmsKeyId?: string;
}
export class Eks extends TerraformModule {
  private readonly inputs: { [name: string]: any } = { }
  public constructor(scope: Construct, id: string, options: EksOptions) {
    super(scope, id, {
      source: 'terraform-aws-modules/eks/aws',
      version: '7.0.1',
    });
    this.attachWorkerCniPolicy = options.attachWorkerCniPolicy;
    this.workersRoleName = options.workersRoleName;
    this.workersAdditionalPolicies = options.workersAdditionalPolicies;
    this.workerSecurityGroupId = options.workerSecurityGroupId;
    this.mapUsers = options.mapUsers;
    this.mapRoles = options.mapRoles;
    this.writeAwsAuthConfig = options.writeAwsAuthConfig;
    this.manageWorkerIamResources = options.manageWorkerIamResources;
    this.clusterEndpointPublicAccess = options.clusterEndpointPublicAccess;
    this.workerCreateSecurityGroup = options.workerCreateSecurityGroup;
    this.kubeconfigAwsAuthenticatorCommandArgs = options.kubeconfigAwsAuthenticatorCommandArgs;
    this.workerSgIngressFromPort = options.workerSgIngressFromPort;
    this.workerAmiOwnerIdWindows = options.workerAmiOwnerIdWindows;
    this.workerAmiNameFilter = options.workerAmiNameFilter;
    this.clusterSecurityGroupId = options.clusterSecurityGroupId;
    this.attachWorkerAutoscalingPolicy = options.attachWorkerAutoscalingPolicy;
    this.kubeconfigAwsAuthenticatorCommand = options.kubeconfigAwsAuthenticatorCommand;
    this.workerAmiNameFilterWindows = options.workerAmiNameFilterWindows;
    this.workersGroupDefaults = options.workersGroupDefaults;
    this.workerGroups = options.workerGroups;
    this.tags = options.tags;
    this.manageClusterIamResources = options.manageClusterIamResources;
    this.workerCreateInitialLifecycleHooks = options.workerCreateInitialLifecycleHooks;
    this.clusterDeleteTimeout = options.clusterDeleteTimeout;
    this.kubeconfigAwsAuthenticatorAdditionalArgs = options.kubeconfigAwsAuthenticatorAdditionalArgs;
    this.workerAmiOwnerId = options.workerAmiOwnerId;
    this.workerGroupsLaunchTemplate = options.workerGroupsLaunchTemplate;
    this.permissionsBoundary = options.permissionsBoundary;
    this.manageWorkerAutoscalingPolicy = options.manageWorkerAutoscalingPolicy;
    this.clusterIamRoleName = options.clusterIamRoleName;
    this.clusterEndpointPrivateAccess = options.clusterEndpointPrivateAccess;
    this.clusterCreateSecurityGroup = options.clusterCreateSecurityGroup;
    this.vpcId = options.vpcId;
    this.subnets = options.subnets;
    this.writeKubeconfig = options.writeKubeconfig;
    this.clusterVersion = options.clusterVersion;
    this.clusterEnabledLogTypes = options.clusterEnabledLogTypes;
    this.clusterCreateTimeout = options.clusterCreateTimeout;
    this.kubeconfigName = options.kubeconfigName;
    this.kubeconfigAwsAuthenticatorEnvVariables = options.kubeconfigAwsAuthenticatorEnvVariables;
    this.iamPath = options.iamPath;
    this.localExecInterpreter = options.localExecInterpreter;
    this.workerAdditionalSecurityGroupIds = options.workerAdditionalSecurityGroupIds;
    this.mapAccounts = options.mapAccounts;
    this.clusterName = options.clusterName;
    this.manageAwsAuth = options.manageAwsAuth;
    this.configOutputPath = options.configOutputPath;
    this.clusterLogRetentionInDays = options.clusterLogRetentionInDays;
    this.clusterLogKmsKeyId = options.clusterLogKmsKeyId;
  }
  public get attachWorkerCniPolicy(): boolean | undefined {
    return this.inputs['attach_worker_cni_policy'] as boolean | undefined;
  }
  public set attachWorkerCniPolicy(value: boolean | undefined) {
    this.inputs['attach_worker_cni_policy'] = value;
  }
  public get workersRoleName(): string | undefined {
    return this.inputs['workers_role_name'] as string | undefined;
  }
  public set workersRoleName(value: string | undefined) {
    this.inputs['workers_role_name'] = value;
  }
  public get workersAdditionalPolicies(): string[] | undefined {
    return this.inputs['workers_additional_policies'] as string[] | undefined;
  }
  public set workersAdditionalPolicies(value: string[] | undefined) {
    this.inputs['workers_additional_policies'] = value;
  }
  public get workerSecurityGroupId(): string | undefined {
    return this.inputs['worker_security_group_id'] as string | undefined;
  }
  public set workerSecurityGroupId(value: string | undefined) {
    this.inputs['worker_security_group_id'] = value;
  }
  public get mapUsers(): any[] | undefined {
    return this.inputs['map_users'] as any[] | undefined;
  }
  public set mapUsers(value: any[] | undefined) {
    this.inputs['map_users'] = value;
  }
  public get mapRoles(): any[] | undefined {
    return this.inputs['map_roles'] as any[] | undefined;
  }
  public set mapRoles(value: any[] | undefined) {
    this.inputs['map_roles'] = value;
  }
  public get writeAwsAuthConfig(): boolean | undefined {
    return this.inputs['write_aws_auth_config'] as boolean | undefined;
  }
  public set writeAwsAuthConfig(value: boolean | undefined) {
    this.inputs['write_aws_auth_config'] = value;
  }
  public get manageWorkerIamResources(): boolean | undefined {
    return this.inputs['manage_worker_iam_resources'] as boolean | undefined;
  }
  public set manageWorkerIamResources(value: boolean | undefined) {
    this.inputs['manage_worker_iam_resources'] = value;
  }
  public get clusterEndpointPublicAccess(): boolean | undefined {
    return this.inputs['cluster_endpoint_public_access'] as boolean | undefined;
  }
  public set clusterEndpointPublicAccess(value: boolean | undefined) {
    this.inputs['cluster_endpoint_public_access'] = value;
  }
  public get workerCreateSecurityGroup(): boolean | undefined {
    return this.inputs['worker_create_security_group'] as boolean | undefined;
  }
  public set workerCreateSecurityGroup(value: boolean | undefined) {
    this.inputs['worker_create_security_group'] = value;
  }
  public get kubeconfigAwsAuthenticatorCommandArgs(): string[] | undefined {
    return this.inputs['kubeconfig_aws_authenticator_command_args'] as string[] | undefined;
  }
  public set kubeconfigAwsAuthenticatorCommandArgs(value: string[] | undefined) {
    this.inputs['kubeconfig_aws_authenticator_command_args'] = value;
  }
  public get workerSgIngressFromPort(): number | undefined {
    return this.inputs['worker_sg_ingress_from_port'] as number | undefined;
  }
  public set workerSgIngressFromPort(value: number | undefined) {
    this.inputs['worker_sg_ingress_from_port'] = value;
  }
  public get workerAmiOwnerIdWindows(): string | undefined {
    return this.inputs['worker_ami_owner_id_windows'] as string | undefined;
  }
  public set workerAmiOwnerIdWindows(value: string | undefined) {
    this.inputs['worker_ami_owner_id_windows'] = value;
  }
  public get workerAmiNameFilter(): string | undefined {
    return this.inputs['worker_ami_name_filter'] as string | undefined;
  }
  public set workerAmiNameFilter(value: string | undefined) {
    this.inputs['worker_ami_name_filter'] = value;
  }
  public get clusterSecurityGroupId(): string | undefined {
    return this.inputs['cluster_security_group_id'] as string | undefined;
  }
  public set clusterSecurityGroupId(value: string | undefined) {
    this.inputs['cluster_security_group_id'] = value;
  }
  public get attachWorkerAutoscalingPolicy(): boolean | undefined {
    return this.inputs['attach_worker_autoscaling_policy'] as boolean | undefined;
  }
  public set attachWorkerAutoscalingPolicy(value: boolean | undefined) {
    this.inputs['attach_worker_autoscaling_policy'] = value;
  }
  public get kubeconfigAwsAuthenticatorCommand(): string | undefined {
    return this.inputs['kubeconfig_aws_authenticator_command'] as string | undefined;
  }
  public set kubeconfigAwsAuthenticatorCommand(value: string | undefined) {
    this.inputs['kubeconfig_aws_authenticator_command'] = value;
  }
  public get workerAmiNameFilterWindows(): string | undefined {
    return this.inputs['worker_ami_name_filter_windows'] as string | undefined;
  }
  public set workerAmiNameFilterWindows(value: string | undefined) {
    this.inputs['worker_ami_name_filter_windows'] = value;
  }
  public get workersGroupDefaults(): any | undefined {
    return this.inputs['workers_group_defaults'] as any | undefined;
  }
  public set workersGroupDefaults(value: any | undefined) {
    this.inputs['workers_group_defaults'] = value;
  }
  public get workerGroups(): any | undefined {
    return this.inputs['worker_groups'] as any | undefined;
  }
  public set workerGroups(value: any | undefined) {
    this.inputs['worker_groups'] = value;
  }
  public get tags(): { [key: string]: string } | undefined {
    return this.inputs['tags'] as { [key: string]: string } | undefined;
  }
  public set tags(value: { [key: string]: string } | undefined) {
    this.inputs['tags'] = value;
  }
  public get manageClusterIamResources(): boolean | undefined {
    return this.inputs['manage_cluster_iam_resources'] as boolean | undefined;
  }
  public set manageClusterIamResources(value: boolean | undefined) {
    this.inputs['manage_cluster_iam_resources'] = value;
  }
  public get workerCreateInitialLifecycleHooks(): boolean | undefined {
    return this.inputs['worker_create_initial_lifecycle_hooks'] as boolean | undefined;
  }
  public set workerCreateInitialLifecycleHooks(value: boolean | undefined) {
    this.inputs['worker_create_initial_lifecycle_hooks'] = value;
  }
  public get clusterDeleteTimeout(): string | undefined {
    return this.inputs['cluster_delete_timeout'] as string | undefined;
  }
  public set clusterDeleteTimeout(value: string | undefined) {
    this.inputs['cluster_delete_timeout'] = value;
  }
  public get kubeconfigAwsAuthenticatorAdditionalArgs(): string[] | undefined {
    return this.inputs['kubeconfig_aws_authenticator_additional_args'] as string[] | undefined;
  }
  public set kubeconfigAwsAuthenticatorAdditionalArgs(value: string[] | undefined) {
    this.inputs['kubeconfig_aws_authenticator_additional_args'] = value;
  }
  public get workerAmiOwnerId(): string | undefined {
    return this.inputs['worker_ami_owner_id'] as string | undefined;
  }
  public set workerAmiOwnerId(value: string | undefined) {
    this.inputs['worker_ami_owner_id'] = value;
  }
  public get workerGroupsLaunchTemplate(): any | undefined {
    return this.inputs['worker_groups_launch_template'] as any | undefined;
  }
  public set workerGroupsLaunchTemplate(value: any | undefined) {
    this.inputs['worker_groups_launch_template'] = value;
  }
  public get permissionsBoundary(): string | undefined {
    return this.inputs['permissions_boundary'] as string | undefined;
  }
  public set permissionsBoundary(value: string | undefined) {
    this.inputs['permissions_boundary'] = value;
  }
  public get manageWorkerAutoscalingPolicy(): boolean | undefined {
    return this.inputs['manage_worker_autoscaling_policy'] as boolean | undefined;
  }
  public set manageWorkerAutoscalingPolicy(value: boolean | undefined) {
    this.inputs['manage_worker_autoscaling_policy'] = value;
  }
  public get clusterIamRoleName(): string | undefined {
    return this.inputs['cluster_iam_role_name'] as string | undefined;
  }
  public set clusterIamRoleName(value: string | undefined) {
    this.inputs['cluster_iam_role_name'] = value;
  }
  public get clusterEndpointPrivateAccess(): boolean | undefined {
    return this.inputs['cluster_endpoint_private_access'] as boolean | undefined;
  }
  public set clusterEndpointPrivateAccess(value: boolean | undefined) {
    this.inputs['cluster_endpoint_private_access'] = value;
  }
  public get clusterCreateSecurityGroup(): boolean | undefined {
    return this.inputs['cluster_create_security_group'] as boolean | undefined;
  }
  public set clusterCreateSecurityGroup(value: boolean | undefined) {
    this.inputs['cluster_create_security_group'] = value;
  }
  public get vpcId(): string | undefined {
    return this.inputs['vpc_id'] as string | undefined;
  }
  public set vpcId(value: string | undefined) {
    this.inputs['vpc_id'] = value;
  }
  public get subnets(): string[] | undefined {
    return this.inputs['subnets'] as string[] | undefined;
  }
  public set subnets(value: string[] | undefined) {
    this.inputs['subnets'] = value;
  }
  public get writeKubeconfig(): boolean | undefined {
    return this.inputs['write_kubeconfig'] as boolean | undefined;
  }
  public set writeKubeconfig(value: boolean | undefined) {
    this.inputs['write_kubeconfig'] = value;
  }
  public get clusterVersion(): string | undefined {
    return this.inputs['cluster_version'] as string | undefined;
  }
  public set clusterVersion(value: string | undefined) {
    this.inputs['cluster_version'] = value;
  }
  public get clusterEnabledLogTypes(): string[] | undefined {
    return this.inputs['cluster_enabled_log_types'] as string[] | undefined;
  }
  public set clusterEnabledLogTypes(value: string[] | undefined) {
    this.inputs['cluster_enabled_log_types'] = value;
  }
  public get clusterCreateTimeout(): string | undefined {
    return this.inputs['cluster_create_timeout'] as string | undefined;
  }
  public set clusterCreateTimeout(value: string | undefined) {
    this.inputs['cluster_create_timeout'] = value;
  }
  public get kubeconfigName(): string | undefined {
    return this.inputs['kubeconfig_name'] as string | undefined;
  }
  public set kubeconfigName(value: string | undefined) {
    this.inputs['kubeconfig_name'] = value;
  }
  public get kubeconfigAwsAuthenticatorEnvVariables(): { [key: string]: string } | undefined {
    return this.inputs['kubeconfig_aws_authenticator_env_variables'] as { [key: string]: string } | undefined;
  }
  public set kubeconfigAwsAuthenticatorEnvVariables(value: { [key: string]: string } | undefined) {
    this.inputs['kubeconfig_aws_authenticator_env_variables'] = value;
  }
  public get iamPath(): string | undefined {
    return this.inputs['iam_path'] as string | undefined;
  }
  public set iamPath(value: string | undefined) {
    this.inputs['iam_path'] = value;
  }
  public get localExecInterpreter(): string[] | undefined {
    return this.inputs['local_exec_interpreter'] as string[] | undefined;
  }
  public set localExecInterpreter(value: string[] | undefined) {
    this.inputs['local_exec_interpreter'] = value;
  }
  public get workerAdditionalSecurityGroupIds(): string[] | undefined {
    return this.inputs['worker_additional_security_group_ids'] as string[] | undefined;
  }
  public set workerAdditionalSecurityGroupIds(value: string[] | undefined) {
    this.inputs['worker_additional_security_group_ids'] = value;
  }
  public get mapAccounts(): string[] | undefined {
    return this.inputs['map_accounts'] as string[] | undefined;
  }
  public set mapAccounts(value: string[] | undefined) {
    this.inputs['map_accounts'] = value;
  }
  public get clusterName(): string | undefined {
    return this.inputs['cluster_name'] as string | undefined;
  }
  public set clusterName(value: string | undefined) {
    this.inputs['cluster_name'] = value;
  }
  public get manageAwsAuth(): string | undefined {
    return this.inputs['manage_aws_auth'] as string | undefined;
  }
  public set manageAwsAuth(value: string | undefined) {
    this.inputs['manage_aws_auth'] = value;
  }
  public get configOutputPath(): string | undefined {
    return this.inputs['config_output_path'] as string | undefined;
  }
  public set configOutputPath(value: string | undefined) {
    this.inputs['config_output_path'] = value;
  }
  public get clusterLogRetentionInDays(): number | undefined {
    return this.inputs['cluster_log_retention_in_days'] as number | undefined;
  }
  public set clusterLogRetentionInDays(value: number | undefined) {
    this.inputs['cluster_log_retention_in_days'] = value;
  }
  public get clusterLogKmsKeyId(): string | undefined {
    return this.inputs['cluster_log_kms_key_id'] as string | undefined;
  }
  public set clusterLogKmsKeyId(value: string | undefined) {
    this.inputs['cluster_log_kms_key_id'] = value;
  }
  public get clusterCertificateAuthorityDataOutput(): string {
    return this.interpolationForOutput('cluster_certificate_authority_data')
  }
  public get workersAsgNamesOutput(): string {
    return this.interpolationForOutput('workers_asg_names')
  }
  public get workersAsgArnsOutput(): string {
    return this.interpolationForOutput('workers_asg_arns')
  }
  public get clusterArnOutput(): string {
    return this.interpolationForOutput('cluster_arn')
  }
  public get workerIamInstanceProfileNamesOutput(): string {
    return this.interpolationForOutput('worker_iam_instance_profile_names')
  }
  public get workersLaunchTemplateLatestVersionsOutput(): string {
    return this.interpolationForOutput('workers_launch_template_latest_versions')
  }
  public get workerAutoscalingPolicyNameOutput(): string {
    return this.interpolationForOutput('worker_autoscaling_policy_name')
  }
  public get clusterSecurityGroupIdOutput(): string {
    return this.interpolationForOutput('cluster_security_group_id')
  }
  public get clusterVersionOutput(): string {
    return this.interpolationForOutput('cluster_version')
  }
  public get clusterEndpointOutput(): string {
    return this.interpolationForOutput('cluster_endpoint')
  }
  public get clusterIdOutput(): string {
    return this.interpolationForOutput('cluster_id')
  }
  public get workerIamInstanceProfileArnsOutput(): string {
    return this.interpolationForOutput('worker_iam_instance_profile_arns')
  }
  public get workersLaunchTemplateArnsOutput(): string {
    return this.interpolationForOutput('workers_launch_template_arns')
  }
  public get workerIamRoleNameOutput(): string {
    return this.interpolationForOutput('worker_iam_role_name')
  }
  public get workersUserDataOutput(): string {
    return this.interpolationForOutput('workers_user_data')
  }
  public get cloudwatchLogGroupNameOutput(): string {
    return this.interpolationForOutput('cloudwatch_log_group_name')
  }
  public get configMapAwsAuthOutput(): string {
    return this.interpolationForOutput('config_map_aws_auth')
  }
  public get workerAutoscalingPolicyArnOutput(): string {
    return this.interpolationForOutput('worker_autoscaling_policy_arn')
  }
  public get workerIamRoleArnOutput(): string {
    return this.interpolationForOutput('worker_iam_role_arn')
  }
  public get workersLaunchTemplateIdsOutput(): string {
    return this.interpolationForOutput('workers_launch_template_ids')
  }
  public get workersDefaultAmiIdOutput(): string {
    return this.interpolationForOutput('workers_default_ami_id')
  }
  public get kubeconfigOutput(): string {
    return this.interpolationForOutput('kubeconfig')
  }
  public get clusterOidcIssuerUrlOutput(): string {
    return this.interpolationForOutput('cluster_oidc_issuer_url')
  }
  public get clusterIamRoleArnOutput(): string {
    return this.interpolationForOutput('cluster_iam_role_arn')
  }
  public get workerSecurityGroupIdOutput(): string {
    return this.interpolationForOutput('worker_security_group_id')
  }
  public get clusterIamRoleNameOutput(): string {
    return this.interpolationForOutput('cluster_iam_role_name')
  }
  public get kubeconfigFilenameOutput(): string {
    return this.interpolationForOutput('kubeconfig_filename')
  }
  protected synthesizeAttributes() {
    return this.inputs;
  }
}
"
`;
